utils/calendrical_calculations/src/hebrew_keviyah.rs

```diff
@@ -0,0 +1,539 @@
+/// are leap years.
+///
+/// The Hebrew calendar further categorizes regular years as whether they come before/after/or
+/// between leap years, and this is used when performing lookups.
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum MetonicCycleType {
+    /// Before a leap year (2, 5, 10, 13, 16)
+    LMinusOne,
+    /// After a leap year (1, 4, 9, 12, 15)
+    LPlusOne,
+    /// Between leap years (7. 18)
+    LPlusMinusOne,
+    /// Leap year (3, 6, 8, 11, 14, 17, 19)
+    Leap,
+}
+
+impl MetonicCycleType {
+    fn for_h_year(h_year: i32) -> Self {
+        // The -1 is because h_year is 1-indexed
+        // The +1 is because our match statement is also 1-indexed
+        // and we want to have this match statement match resources that list
+        // these year types (both Adjler and Wikipedia).
+        match ((h_year - 1) % 19) + 1 {
+            2 | 5 | 10 | 13 | 16 => Self::LMinusOne,
+            1 | 4 | 9 | 12 | 15 => Self::LPlusOne,
+            7 | 18 => Self::LPlusMinusOne,
+            _ => Self::Leap,
+        }
+    }
+}
+
+// The actual Four Gates tables.
+//
+// Each entry is a range (ending at the next entry), and it corresponds to the equivalent discriminant value of the Keviyah type.
+// Leap and regular years map to different Keviyah values, however regular years all map to the same set of
+// seven values, with differing ḥalakim bounds for each. The first entry in the Four Gates table straddles the end of the previous week
+// and the beginning of this one.
+//
+// The regular-year tables only differ by their third and last entries (We may be able to write this as more compact code)
+//
+// You can reference these tables from https://en.wikipedia.org/wiki/Hebrew_calendar#The_four_gates
+// or from Adjler (Appendix 4). Be sure to look at the Adjler table referring the "modern calendar", older tables
+// use slightly different numbers.
+const FOUR_GATES_LMINUSONE: [i32; 7] = [
+    ḥal!(7 - 18 - 0),
+    ḥal!(1 - 9 - 204),
+    ḥal!(2 - 18 - 0),
+    ḥal!(3 - 9 - 204),
+    ḥal!(5 - 9 - 204),
+    ḥal!(5 - 18 - 0),
+    ḥal!(6 - 9 - 204),
+];
+const FOUR_GATES_LPLUSONE: [i32; 7] = [
+    ḥal!(7 - 18 - 0),
+    ḥal!(1 - 9 - 204),
+    ḥal!(2 - 15 - 589),
+    ḥal!(3 - 9 - 204),
+    ḥal!(5 - 9 - 204),
+    ḥal!(5 - 18 - 0),
+    ḥal!(6 - 0 - 408),
+];
+
+const FOUR_GATES_LPLUSMINUSONE: [i32; 7] = [
+    ḥal!(7 - 18 - 0),
+    ḥal!(1 - 9 - 204),
+    ḥal!(2 - 15 - 589),
+    ḥal!(3 - 9 - 204),
+    ḥal!(5 - 9 - 204),
+    ḥal!(5 - 18 - 0),
+    ḥal!(6 - 9 - 204),
+];
+
+const FOUR_GATES_LEAP: [i32; 7] = [
+    ḥal!(7 - 18 - 0),
+    ḥal!(1 - 20 - 491),
+    ḥal!(2 - 18 - 0),
+    ḥal!(3 - 18 - 0),
+    ḥal!(4 - 11 - 695),
+    ḥal!(5 - 18 - 0),
+    ḥal!(6 - 20 - 491),
+];
+
+/// Perform the four gates calculation, giving you the Keviyah for a given year type and
+/// the ḥalakim-since-beginning-of-week of its molad Tishri
+fn keviyah_for(year_type: MetonicCycleType, ḥalakim: i32) -> Keviyah {
```

Comment:
"year type" was previously defined to mean Deficient/Regular/Complete

```suggestion
/// Perform the four gates calculation, giving you the Keviyah for a given metonic type and
/// the ḥalakim-since-beginning-of-week of its molad Tishri
fn keviyah_for(metonic_type: MetonicCycleType, ḥalakim_of_tishri: i32) -> Keviyah {
```